client/asset/eth/eth.go

```diff
@@ -313,28 +318,151 @@ func (*ExchangeWallet) PreRedeem(req *asset.PreRedeemForm) (*asset.PreRedeem, er
 	return nil, asset.ErrNotImplemented
 }
 
+// EthCoin implements the asset.Coin interface for ETH
```

Comment:
Wouldn't the compiler already check this wherever it is being used? For example when it's being returned in `FundOrder`.

Reply:
In this case it's ensured by `FundOrder`, yes, but that could change.  If it is intended to satisfy a certain interface, especially if the type is exported like `EthCoin` is, it is valuable to have a statement like `var _ asset.Coin = (*EthCoin)(nil)` (or similar if it should work with a value receiver), if only to document the intent.

Reply:
BTW, why is `EthCoin` exported?  That required for future work?

Reply:
Ok that makes sense. 

Now that I think about it, this might need to be moved to another location, maybe into the `dex/networks/eth` code. @JoeGruffins you're currently working on the `FundingCoin` method in the eth server code. You will need to parse these Coin IDs, correct?

Reply:
For balance was it? Let me look around. In my mind coin ID is something else though. This would be a balance id?

Reply:
You're right about funding order. But here's an example of a situation you can avoid by checking. Remove the check and it gets by the compiler and panics, I bleieve. I suppose the call site could check better as well though. https://play.golang.org/p/zUEspd6olwP

Reply:
So, eth coinID.

I had intended the coin ID to be defined here https://github.com/decred/dcrdex/blob/master/server/asset/eth/common.go

Currently we have two types. One represents a transaction. It is useful for a swap before it is mined as can be seen in the server's Contract pr. Then we have a swap type which should be useful after the tx is mined and the swap is live, as it points to the secret hash and contract. I guess we need a third type for amounts? I would imagine it would be the same as the coin ID you have here, with just the account's public address and the amount of gwei, and it would be the next coinID type (3)

Reply:
> You're right about funding order. But here's an example of a situation you can avoid by checking. Remove the check and it gets by the compiler and panics, I bleieve. I suppose the call site could check better as well though. https://play.golang.org/p/zUEspd6olwP

Are there cases where you would want to cast a type like that?  I think it makes sense for documentation's sake, but other than that I can't think of another reason, maybe I'm missing something though. I think the reason make a type adhere to an interface would be to pass it into a function who's parameter must implement an interface, or to return it from a function who's return type must implement an interface. In both of these cases the compiler would catch the issue.

Reply:
> So, eth coinID.
> 
> I had intended the coin ID to be defined here https://github.com/decred/dcrdex/blob/master/server/asset/eth/common.go
> 
> Currently we have two types. One represents a transaction. It is useful for a swap before it is mined as can be seen in the server's Contract pr. Then we have a swap type which should be useful after the tx is mined and the swap is live, as it points to the secret hash and contract. I guess we need a third type for amounts? I would imagine it would be the same as the coin ID you have here, with just the account's public address and the amount of gwei, and it would be the next coinID type (2)

Yeah I think it makes sense to put these together.

Reply:
> . I think the reason make a type adhere to an interface would be to pass it into a function who's parameter must implement an interface, or to return it from a function who's return type must implement an interface. In both of these cases the compiler would catch the issue.

This is primarily good to have for when package A implements a concrete type that should satisfy an interface defined in package B.  This asset backend happens to be returning the concrete type in the targeted interface anyway, but that's the only reason why compiler checks this, and it doesn't have to be that way.
The explicit check doesn't hurt.

Consider the postgresql driver github.com/lib/pq: https://github.com/lib/pq/commit/730d555dd5aa0f0f5bd365036f148b5dfce03e79
They didn't even realize they had not correctly implemented `database/sql/driver.Driver`.  They fixed it and added this kind of compile time check to make sure that never breaks.  Although here we have something returning the concrete inside the targeted interface already, thus making this check redundant, it could have come together differently.  The usual mantra is actually "accept interfaces, return structs" and had we done that we'd require these checks, otherwise only a consumer of this package would find out if it satisfies the interface, not the package itself.

We do this in a number of places in dcrdex.  One example is the implementation of the mt19937 prng: https://github.com/decred/dcrdex/blob/8bd3a34054f6961493cae64e8958a7e7d90e06d9/server/matcher/mt19937/mt19937_test.go#L15-L17
The package is designed to provide a `rand.Source` and `rand.Source64`, but there's no code *in that package* that would ensure that other than these compile time checks.

Reply:
Cool, thanks for the explanation @chappjc

Reply:
Since now we will have 3 different types of coin IDs that all contain completely different information, I've done a bit of refactoring. Let me know what you think.